[inductor] Fix torch.cond stride mismatch when subgraph outputs are padded

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #169963

Summary: Fixes https://github.com/pytorch/pytorch/issues/169778. Because Inductor can add padding to a subgraph output, we need to force subgraph out strides to match the targets so that both torch.cond branches will produce the same tensor shapes.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

desertfire committed:
[inductor] Fix torch.cond stride mismatch when subgraph outputs are padded

Summary: Fixes https://github.com/pytorch/pytorch/issues/169778.

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Fix torch.cond stride mismatch when subgraph outputs are padded"


Summary: Fixes https://github.com/pytorch/pytorch/issues/169778.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo mlazos

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Fix torch.cond stride mismatch when subgraph outputs are padded"


Summary: Fixes https://github.com/pytorch/pytorch/issues/169778. Because Inductor can add padding to a subgraph output, we need to force the subgraph stride to match the target so that both torch.cond branches will produce the same tensor shapes.

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Fix torch.cond stride mismatch when subgraph outputs are padded"


Summary: Fixes https://github.com/pytorch/pytorch/issues/169778. Because Inductor can add padding to a subgraph output, we need to force subgraph out strides to match the targets so that both torch.cond branches will produce the same tensor shapes.

[ghstack-poisoned]

eellison reviewed (approved):
Looks good, please update require_exact_strides to handle symints, i dont think it does today.

desertfire committed:
Update on "[inductor] Fix torch.cond stride mismatch when subgraph outputs are padded"


Summary: Fixes https://github.com/pytorch/pytorch/issues/169778. Because Inductor can add padding to a subgraph output, we need to force subgraph out strides to match the targets so that both torch.cond branches will produce the same tensor shapes.

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge

desertfire commented on torch/_inductor/ir.py:
```diff
@@ -8810,6 +8821,27 @@ def create(
+                    fake_strides = [
+                        s.node.expr if isinstance(s, torch.SymInt) else s
+                        for s in fake.stride()
+                    ]
```
I thought about that, but there are multiple places need to be refactored. Let me do that in another PR.

desertfire commented on torch/_inductor/ir.py:
```diff
@@ -8810,6 +8821,27 @@ def create(
         fake_operands = [cast(Node, x).meta["val"] for x in fx_operands]
+        fake_outputs = V.graph.current_node.meta["val"]
+
+        def _require_exact_strides(
```
I will do it separately.
